scalameta/common/shared/src/main/scala/scala/meta/internal/trees/NamerMacros.scala

```diff
@@ -34,46 +40,66 @@ trait CommonNamerMacros extends MacroHelpers {
     classifierBoilerplate
   }
 
+  private val quasiTypeName = TypeName(CommonNamerMacros.quasiName)
+  def isQuasiClass(cdef: ClassDef) = cdef.name.toString == CommonNamerMacros.quasiName
+
   def mkQuasi(
       name: TypeName,
       parents: List[Tree],
       paramss: List[List[ValDef]],
-      extraAbtractDefs: List[Tree],
+      extraAbstractDefs: List[Tree],
       extraStubs: String*
   ): ClassDef = {
     val qmods = Modifiers(NoFlags, TypeName("meta"), List(q"new $AstAnnotation"))
-    val qname = TypeName("Quasi")
+    val qname = quasiTypeName
     val qparents = tq"$name" +: tq"$QuasiClass" +: parents.map({
-      case Ident(name) => Select(Ident(name.toTermName), TypeName("Quasi"))
-      case Select(qual, name) => Select(Select(qual, name.toTermName), TypeName("Quasi"))
+      case Ident(name) => Select(Ident(name.toTermName), quasiTypeName)
+      case Select(qual, name) => Select(Select(qual, name.toTermName), quasiTypeName)
       case unsupported => c.abort(unsupported.pos, "implementation restriction: unsupported parent")
     })
 
-    val qstats = ListBuffer[Tree]()
+    val qstats = mutable.ListBuffer[Tree]()
     qstats += q"""
       def pt: _root_.java.lang.Class[_] = {
         $ArrayClassMethod($ClassOfMethod[$name], this.rank)
       }
     """
 
-    def stub() = {
+    val stub = {
       val unsupportedUnquotingPosition = "unsupported unquoting position"
       val unsupportedSplicingPosition = "unsupported splicing position"
       val message =
         q"if (this.rank == 0) $unsupportedUnquotingPosition else $unsupportedSplicingPosition"
       q"throw new $UnsupportedOperationException($message)"
     }
-    val qstubs = (paramss.flatten.map(_.name.toString) ++ extraStubs).distinct.map(TermName.apply)
-    qstubs.foreach(name => qstats += q"def $name: $NothingClass = ${stub()}")
+    val stubbedMembers = new mutable.HashSet[String]
```

Comment:
Why is `stubbedMembers` needed? I see that it's never really read.

Reply:
to make sure that, for instance, `value` is not defined twice, since `extraStubs` are hardcoded to contain it. 

in the original code, `distinct` was used to remove duplicates between class parameters and extra stubs. but we also need to apply the same to extra abstract definitions as well.